### PR TITLE
Feat/impl localcache

### PR DESCRIPTION
**Pull request by RayCh8**
* Implement localcache
* Go training

### CODE REVIEW

File: localcache.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	// Time to live of cache key.
+	DefaultTTL = time.Second * 30
+	// Error message means that key does not exist.
```
Comment (VineyAT) [inline localcache.go]: comment starts with the name `ErrKeyNotExist`. e.g. // `ErrKeyNotExist` means key doesn't exist

Comment (VineyAT) [reply localcache.go]: Could you please fix all the same issues as this?

File: impl.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	e := entry{
+		val:    val,
+		expiry: time.Now().Add(DefaultTTL),
+	}
+
+	c.es[key] = &e
+}
+
+// New return Cache instance
+func New() cache {
```
Comment (VineyAT) [inline impl.go]: New should return `Cache` interface not `cache` struct

File: localcache.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package localcache
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	// Time to live of cache key.
+	DefaultTTL = time.Second * 30
```
Comment (VineyAT) [inline localcache.go]: Could you explain why this need to be public? or keep it private.

File: impl_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	got := len(c.es)
+
+	assert.Equal(t, got, 0, "New cache size should be zero.")
+}
+
+func TestGet(t *testing.T) {
+	c := New()
+
+	c.Set("a", 123)
+	got, ok := c.Get("a")
```
Comment (VineyAT) [inline impl_test.go]: Actually, we prefer to use the following instead:

```go
val, err := c.Get("a")
```

Comment (VineyAT) [reply impl_test.go]: Could you fix the following cases similar to this naming?

File: impl_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+	got := len(c.es)
+
+	assert.Equal(t, got, 0, "New cache size should be zero.")
+}
+
+func TestGet(t *testing.T) {
+	c := New()
+
+	c.Set("a", 123)
```
Comment (VineyAT) [inline impl_test.go]: You use two method provided by the same interface. It's weird that we are not sure their quality but use both of them to verify each other. Fix this line

File: impl_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	assert.Equal(t, size, 1, "Cache size should be one")
+	assert.Equal(t, got, 123, "Should be equal")
+}
+
+func TestSet(t *testing.T) {
+	c := New()
+
+	c.Set("a", 123)
+	c.Set("a", "456")
+	got, ok := c.Get("a")
```
Comment (VineyAT) [inline impl_test.go]: the same issue mentioned above. fix it

File: impl_test.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+	assert.Equal(t, got, "456", "Should be equal")
+	assert.Nil(t, ok, "Should be nil")
+}
+
+func TestKeyNotExist(t *testing.T) {
```
Comment (VineyAT) [inline impl_test.go]: `TestGet_NotExist`

File: impl_test.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+	assert.Nil(t, got)
+	assert.Equal(t, ok, ErrKeyNotExist, "Key doesn't exist in cache")
+}
+
+func TestKeyExpiry(t *testing.T) {
+	c := New()
+
+	c.Set("a", 123)
+	time.Sleep(30 * time.Second)
```
Comment (VineyAT) [inline impl_test.go]: It is not good. The CI will wait until timeout. Please use fake data instead.

File: impl.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	val    interface{}
+	expiry time.Time
+}
+
+type cache struct {
+	es map[string]*entry
+}
+
+func (c *cache) Get(key string) (interface{}, error) {
+	e, ok := c.es[key]
```
Comment (VineyAT) [inline impl.go:27]: If concurrent access this struct, what will happen? Could you please prevent this case?

File: localcache.go (hunk 1 above)
Comment (RayCh8) [reply localcache.go]: ok

File: impl_test.go (hunk 4 above)
Comment (RayCh8) [reply impl_test.go]: ok

File: impl_test.go (hunk 8 above)
Comment (RayCh8) [reply impl_test.go]: ok

File: impl.go (hunk 9 above)
Comment (RayCh8) [reply impl.go:27]: If simultaneously write value to same key, it possible to get different value from same key.
I will add mutex to lock write and read action.

File: impl_test.go (hunk 7 above)
Comment (RayCh8) [reply impl_test.go]: Would it be better to rename it to `TestGet_KeyNotExist`?

File: impl.go (hunk 2 above)
Comment (RayCh8) [reply impl.go]: ok

File: localcache.go (hunk 3 above)
Comment (RayCh8) [reply localcache.go]: My initial thought was that the TTL could be decided by the user. But maybe it's better to use a public method to change the TTL time.

File: impl_test.go (hunk 5 above)
Comment (RayCh8) [reply impl_test.go]: ok

File: localcache.go [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"errors"
+	"time"
+)
+
+var (
+	defaultTTL = time.Second * 30
+	// ErrKeyNotExist means that key doesn't exist.
+	ErrKeyNotExist = errors.New("key does not exist.")
+	// ErrKeyExpiry means key is expiried.
```
Comment (VineyAT) [inline localcache.go:12]: typo `expired`

File: localcache.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	defaultTTL = time.Second * 30
+	// ErrKeyNotExist means that key doesn't exist.
+	ErrKeyNotExist = errors.New("key does not exist.")
```
Comment (VineyAT) [inline localcache.go:11]: A small suggestion: move these two Error messages from impl.go to localcache.go. It's a public error need to be known outside.

File: impl_test.go (hunk 7 above)
Comment (VineyAT) [reply impl_test.go]: Of course. the suffix naming depends on your use case.

**Review by VineyAT — APPROVED** [review]
ac with comments.

**Pull request merged**